### ISSUE TITLE
128-bit CAS without Assembly Broken?

### ISSUE THREAD

**Issue by nmhamster** [kokkos/kokkos#161]
If I comment out the assembly version of the 128-bit CAS loop and compile with OpenMP tests fail on X86 systems with GNU and Intel. Not sure but the fall-back may be broken which could have impact on other platforms.

**Comment by crtrott** [comment]
Ok I think this was only in pull request #160

The problem is that the use of cas128 was only protected by "if defined(KOKKOS_ENABLE_ASM)" while the contents of cas128 with the changes in the pull request were now protected by
"if defined(KOKKOS_ENABLE_ASM) && defined ( KOKKOS_USE_ISA_X86_64 )".
That way the 128 cas which got called was actually not atomic, which made the tests fail since the cas loops never finished.
I changed it so that the whole function is not visible anymore (not just the loop body) if those two things are not defined, and I protected the use of cas128 with both defines everywhere. 

So now if either KOKKOS_ENABLE_ASM or KOKKOS_USE_ISA_X86_64 is not defined 128 bit atomics go to the generic backend.
